server/auth/types/multiple/multi_auth.ts

```diff
@@ -0,0 +1,293 @@
+    this.openIdAuthConfig.tokenEndpoint = payload.token_endpoint;
+    this.openIdAuthConfig.endSessionEndpoint = payload.end_session_endpoint;
+  }
+
+  async isValidCookie(cookie: SecuritySessionCookie): Promise<boolean> {
+    if (cookie.authType === AuthType.BASIC) {
+      const result =
+        cookie.expiryTime &&
+        ((cookie.username && cookie.credentials?.authHeaderValue) ||
+          (this.config.auth.anonymous_auth_enabled && cookie.isAnonymousAuth));
+
+      return result;
+    } else if (cookie.authType === AuthType.OPEN_ID) {
+      if (
+        !cookie.username ||
+        !cookie.expiryTime ||
+        !cookie.credentials?.authHeaderValue ||
+        !cookie.credentials?.expires_at
+      ) {
+        return false;
+      }
+      if (cookie.credentials?.expires_at > Date.now()) {
+        return true;
+      }
+
+      // need to renew id token
+      if (cookie.credentials.refresh_token) {
+        try {
+          const query: any = {
+            grant_type: 'refresh_token',
+            client_id: this.config.openid?.client_id,
+            client_secret: this.config.openid?.client_secret,
+            refresh_token: cookie.credentials.refresh_token,
+          };
+          const refreshTokenResponse = await callTokenEndpoint(
+            this.openIdAuthConfig.tokenEndpoint!,
+            query,
+            this.wreckClient
+          );
+
+          // if no id_token from refresh token call, maybe the Idp doesn't allow refresh id_token
+          if (refreshTokenResponse.idToken) {
+            cookie.credentials = {
+              authHeaderValue: `Bearer ${refreshTokenResponse.idToken}`,
+              refresh_token: refreshTokenResponse.refreshToken,
+              expires_at: Date.now() + refreshTokenResponse.expiresIn! * 1000, // expiresIn is in second
+            };
+            return true;
+          } else {
+            return false;
+          }
+        } catch (error: any) {
+          this.logger.error(error);
+          return false;
+        }
+      } else {
+        // no refresh token, and current token is expired
+        return false;
+      }
+    } else if (cookie.authType === AuthType.SAML) {
+      return cookie.username && cookie.expiryTime && cookie.credentials?.authHeaderValue;
+    } else {
+      return false;
+    }
+  }
+
+  handleUnauthedRequest(
+    request: OpenSearchDashboardsRequest,
+    response: LifecycleResponseFactory,
+    toolkit: AuthToolkit
+  ): OpenSearchDashboardsResponse {
+    if (this.isPageRequest(request)) {
+      const nextUrlParam = composeNextUrlQueryParam(
+        request,
+        this.coreSetup.http.basePath.serverBasePath
+      );
+
+      return response.redirected({
+        headers: {
+          location: `${this.coreSetup.http.basePath.serverBasePath}${LOGIN_PAGE_URI}?${nextUrlParam}`,
+        },
+      });
+    } else {
+      return response.unauthorized({
+        body: `Authentication required`,
```

Comment:
Should not this respond `401 Unauthorized`?

Reply:
Addressed.